### PR TITLE
[WIP] Create unified framework based on zechengh09/BBR-Receiver

### PR DESCRIPTION
**Pull request by ccanel**
This pull request combines the `mathis-model` and `ai` branches of `zecheng9/BBR-Receiver` in an extensible manner to enable experimentation with different methods of computing the bandwidth fair share and ACK pacing interval.

The entry point is `ns-3-unfair/scratch/ai-multi-flow.cc`. I have temporarily removed other `scratch/` scripts, but will add them back. 

This is still a work in progress. Ignore any `*.py` files for now, and note that I will remove changes that only modify whitespace (I have configured my editor to automatically remove trailing whitespace, which creates extraneous changes --- I will remove these).

### CODE REVIEW

**Comment by ccanel** [comment]
Ignore the commit names for now. When we merge this PR, we will squash all of its commits into one.

File: avg.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import json, argparse
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("data", help="Data json file")
+    args = parser.parse_args()
+
+    with open(args.data) as f:
+        data = json.load(f)
+
+    DUR_s = 300
```
Comment (zechengh09) [inline avg.py]: Could you explain what does DUS_s stand for? I'm wondering why less than 300 is considered to be with ack pacing

File: gen_training_data.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+DELAY_DELTA_us = 50
+DELAYS_us = list(range(DELAY_MIN_us, DELAY_MAX_us + 1, DELAY_DELTA_us))
+# Ack period (us).
+ACK_PERIOD_MIN_us = 0
+ACK_PERIOD_MAX_us = 6_000  # 1 RTT ≈ 4_000
+ACK_PERIOD_DELTA_us = 50
+ACK_PERIODS_us = list(range(
+    ACK_PERIOD_MIN_us, ACK_PERIOD_MAX_us + 1, ACK_PERIOD_DELTA_us))
+# Router queue size (packets).
+QUEUE = 100
```
Comment (zechengh09) [inline gen_training_data.py]: Just curious - are we only training data on queue size of 100 packets?

File: parse_pcap.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+
+# Whether to plot per-experiment average throughput.
+DO_PLT = False
+# The number of link segments on each round trip path.
+LINKS_PER_RTT = 4
+# The IP address of the sender.
+SRC_IP = "10.1.1.1"
+# The number of packets to consider when computing instantaneous bandwidth.
+TPT_WINDOW = 5
```
Comment (zechengh09) [inline parse_pcap.py]: I guess we'll update this for the new model

File: parse_pcap.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+    Each item in "pkts" should be a pair of: (PacketMetadata, packet). The
+    return value is a pair of: (timestamp (s), throughput (bps)). The timestamp
+    corresponds to the last packet in pkts.
+
+    The temporal window is defined by the difference in time between the first
+    and last packets. However, when computing the number of bytes of data, the
+    first packet is discarded. This is because I believe that the timestamps are
+    end times, meaning that the timestamp of the first packet is best thought of
+    as the time at which the *second* packet *began* arriving.
```
Comment (zechengh09) [inline parse_pcap.py]: Good point!

File: scratch/ai-multi-flow.cc [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+NS_LOG_COMPONENT_DEFINE ("main");
+
+const int BBR_PRINT_PERIOD = 2;  // sec
+
+std::vector<Ptr<PacketSink>> sinks;
+extern bool useReno;
+
+
+Ptr<PacketSink> CreateFlow(uint16_t port, Ipv4InterfaceContainer i1i2,
+                           NodeContainer nodes, double durS, uint32_t recalcUs)
```
Comment (zechengh09) [inline scratch/ai-multi-flow.cc]: Seems like recalcUs is not used in this function

File: scratch/ai-multi-flow.cc [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+  // Note: BBR still works without this.
+  Config::SetDefault ("ns3::TcpSocket::DelAckCount", UintegerValue (0));
+  // Increase the capacity of the send and receive buffers to make sure that the
+  // experiment is not application-limited.
+  Config::SetDefault ("ns3::TcpSocket::SndBufSize", UintegerValue (1'000'000u));
+  Config::SetDefault ("ns3::TcpSocket::RcvBufSize", UintegerValue (1'000'000u));
+  // Configure TcpSocketBase with the model filepath.
+  Config::SetDefault ("ns3::TcpSocketBase::UnfairMitigationEnable",
+                      BooleanValue (true));
+  Config::SetDefault ("ns3::TcpSocketBase::FairShareEstimationType",
```
Comment (zechengh09) [inline scratch/ai-multi-flow.cc:202]: Thanks for adding this :)

File: src/internet/model/bbr-tag.h [hunk 7]
```diff
@@ -0,0 +1,60 @@
+
+#pragma once
+
+#include "ns3/type-id.h"
+#include "ns3/tcp-header.h"
+#include "ns3/uinteger.h"
+
+namespace ns3 {
+
+class BbrTag : public Tag {
```
Comment (zechengh09) [inline src/internet/model/bbr-tag.h:10]: Is this used for distinguishing BBR or other flows in the test file?

File: src/internet/model/tcp-socket-base.cc [hunk 8]
```diff
@@ -4156,4 +4098,592 @@ RttHistory::RttHistory (const RttHistory& h)
...
 }
 
+
+// ACK pacing
+
+void
+TcpSocketBase::Unfair (Ptr<Packet> p, SequenceNumber32 seq) {
+  NS_LOG_FUNCTION (this << p << seq);
+  // If this is not a BBR flow, then abort.
+  if (! m_receivingBbr)
```
Comment (zechengh09) [inline src/internet/model/tcp-socket-base.cc:4150]: I was under the assumption that ACK pacing would be applied to all flows, regardless of their congestion type.

File: src/internet/model/tcp-socket-base.cc [hunk 9]
```diff
@@ -4156,4 +4098,592 @@ RttHistory::RttHistory (const RttHistory& h)
...
+  if (fairTput == -1)
+    {
+      m_ackPeriod = Seconds (0);
+      return;
+    }
+
+  Stats stats = GetStats ();
+  double actualTput = stats.tputMbps;
+
+  // Do we want to do this?
```
Comment (zechengh09) [inline src/internet/model/tcp-socket-base.cc:4213]: I'm not sure... Would this be a parameter to set during training?

File: src/internet/model/tcp-socket-base.cc [hunk 10]
```diff
@@ -4156,4 +4098,592 @@ RttHistory::RttHistory (const RttHistory& h)
...
+  // Constant C.
+  double c = 1.2247;
+  double loss_rate = loss_count / (1.0 * (loss_count + m_arrivalTimes.size ()));
+  double fairTput = 0;
+  switch (m_fairShareType)
+    {
+    case FairShareEstimationType::Mathis:
+      if (loss_rate > 0)
+        {
+          fairTput = c / (rttS * sqrt (loss_rate)); // In number of packets per second
```
Comment (zechengh09) [inline src/internet/model/tcp-socket-base.cc:4348]: I just realized that this would overestimate the fair throughput. Imagine the loss count of last eight RTTS (latest to earliest) to be (10, 10, 10, 0, 0, 0, 0, 0), the moving average after 6 RTT would be (0, 0, 0, 0, 0, 0, 10, 10). The estimation is even higher after the next RTT. I tested it and removing the zero-loss RTTs at the end would give a more accurate fair share estimate.

I will update this once your code is merged in.

File: src/internet/model/tcp-socket-base.cc [hunk 11]
```diff
@@ -4156,4 +4098,592 @@ RttHistory::RttHistory (const RttHistory& h)
...
+  double out = m_net->forward (std::vector<torch::jit::IValue> ({
+        Scale (targetTputMbps * 1e6,
+               std::get<0> (std::get<0> (m_scaleParams)[0]),
+               std::get<1> (std::get<0> (m_scaleParams)[0]), 0, 1)}))
+    .toTensor ().item ().to<double> ();
+  double out_unscaled = Scale (out, 0, 1,
+                               std::get<0> (std::get<1> (m_scaleParams)[0]),
+                               std::get<1> (std::get<1> (m_scaleParams)[0]));
+  NS_LOG_INFO ("Raw output: " << out << ", unscaled: " << out_unscaled);
+  return MicroSeconds (1. / out_unscaled);
```
Comment (zechengh09) [inline src/internet/model/tcp-socket-base.cc:4409]: Is this assuming that the model would produce the delay time?

File: src/internet/model/tcp-socket-base.cc [hunk 12]
```diff
@@ -4156,4 +4098,592 @@ RttHistory::RttHistory (const RttHistory& h)
...
+                                  Ptr<NetDevice> boundnetdevice)
+{
+  NS_LOG_FUNCTION (this << tcp << p << header << localaddr << peeraddr <<
+                   boundnetdevice);
+  // If the ACK period is 0 or the next time at which an ACK should be sent is
+  // in the past, then send the ACK immediately. Otherwise, enqueue it for
+  // tranmission later.
+  if (m_ackPeriod == 0 || m_prevAckTime + m_ackPeriod < Simulator::Now ())
+    {
+      // TODO: Debug why this assertion is sometimes invalid.
```
Comment (zechengh09) [inline src/internet/model/tcp-socket-base.cc]: I think it's because the measured throughput goes from above 1.2x to lower 1.2x. For the previous packet, ACK pacing is still on and pendingAcks queue is not empty. Once the measured throughput is lower than 1.2x, m_ackPeriod == 0, but there're still packets inside the queue.

**Review by zechengh09 — APPROVED** [review]
That's a lot of work and thanks so much for merging them!
